### DISCUSSION TITLE
How do you handle button presses in a timely manner?

### DISCUSSION THREAD

**Discussion by philippejadin** [pschatzmann/arduino-audio-tools#1510] [Q&A] [answered]
Hi, and sorry in advance if this is not the right place. I'm asking here because I'll get answers related to my specific use case, related to audio-tools :)

So, I use this button library to handle simple buttons connected to a pi pico :   https://github.com/LennartHennigs/Button2
It needs buttons to be polled as often as possible. Since I already have the audio copier in the loop (which takes something like 200 ms to complete), the user interface is a bit slugish because there is a delay created by the audio copier.

I tried using interrupts, but it does not feel right and I'm not even sure it would fix my issue.
The idea is to use buttons to start / stop / pause some audio.

I thought about reducing the audio buffer, but which buffer do I need to decrease? Will it affect audio quality?

I'm playing mp3 files from an SD card to adafruit I2s mono amp.

The project will soon be posted in github and will be open source including hardware. If you need the code to help me, let me know, I'll put the project online. 

Thank you everyone in advance and thank you Phil for the great lib!

**Comment by pschatzmann** [comment, marked as answer]
A couple of infos:
- the audiotools also contains some button handling functionality: just have a look at the Wiki: [Interactive Audio](https://github.com/pschatzmann/arduino-audio-tools/wiki/Interactive-Audio)
- In the [copier](https://pschatzmann.github.io/arduino-audio-tools/classaudio__tools_1_1_stream_copy.html) you can specify the copy size: the default value is 1K. If you decrease the value the loop is called more often
- The RP2040 supports FreeRTOS: have a look at the [Multicore Processing](https://github.com/pschatzmann/arduino-audio-tools/wiki/Multicore-Processing) Wiki

**Reply by philippejadin** [reply]
I solved my problem by resizing the copier size to 512 with :    
`copier.resize(512);`

Simple, and in a way provides forced debouncing of my switches :-)

**Reply by pschatzmann** [reply]
Defining StreamCopy copier(out, in, 512); would have been a bit more effiicent
